File: Largest Rectangle in Histogram - Leetcode 84/Largest Rectangle in Histogram - Leetcode 84.cpp
#include <vector>
#include <stack>
#include <algorithm>

class Solution {
public:
    int largestRectangleArea(std::vector<int>& heights) {
        int n = heights.size();
        std::stack<std::pair<int, int>> stk;
        int maxArea = 0;

        for (int i = 0; i < n; i++) {
            int height = heights[i];
            int start = i;
            while (!stk.empty() && height < stk.top().first) {
                auto [h, j] = stk.top(); stk.pop();
                int w = i - j;
                maxArea = std::max(maxArea, h * w);
                start = j;
            }
            stk.push({height, start});
        }

        while (!stk.empty()) {
            auto [h, j] = stk.top(); stk.pop();
            int w = n - j;
            maxArea = std::max(maxArea, h * w);
        }

        return maxArea;
    }
};

File: Largest Rectangle in Histogram - Leetcode 84/Largest Rectangle in Histogram - Leetcode 84.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        n = len(heights)
        stk = []
        max_area = 0

        for i, height in enumerate(heights):
            start = i
            while stk and height < stk[-1][0]:
                h, j = stk.pop()
                w = i - j
                a = h*w
                max_area = max(max_area, a)
                start = j
            stk.append((height, start))
        
        while stk:
            h, j = stk.pop()
            w = n - j
            max_area = max(max_area, h*w)
        
        return max_area
        # Time: O(n)
        # Space: O(n)